File: 프로그래머스/lv2/42883. 큰 수 만들기/큰 수 만들기.cpp
#include <bits/stdc++.h>

using namespace std;
string maxStr = "";
string solution(string number, int k) {
    string answer = "";
    string original = number;
    int kk = k;
    int startIdx = 0;
    
    while(k && startIdx < number.size()) {
        // cout << "number : " << number << ' ' << "k : " << k << ' ';
        int maxIdx = startIdx;
        char maxChar = number[startIdx];
        int minLen = min((int)number.length(), startIdx+k+1);
        
        for(int i = startIdx + 1 ; i < minLen ; i++) {
            if(maxChar < number[i]) {
                maxIdx = i;
                maxChar = number[i];
            }
        }
        // cout << "startIdx : " << startIdx << ' ' << "maxIdx : " << maxIdx << ' ';
        // cout << "maxChar : " << maxChar << '\n';
        if(maxIdx == startIdx){
            startIdx++;
        } else {
            number.erase(startIdx, maxIdx-startIdx);
            k -= (maxIdx-startIdx);
        }       
    }
    
    if(original == number) {
        // cout << number << '\n';
        number.erase(number.size()-kk, kk);
    }
    answer += number;
    return answer;
}

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.js
function solution(number, k) {
    var answer = '';
    let str = number.split('');
    let idx = 0;
    while(k && idx < str.length) {
        let maxIdx = idx;
        let maxNum = str[idx];
        let minLen = Math.min(str.length, idx + k + 1);
        
        for(let i = idx + 1; i < minLen ; i++) {
            if(str[i] > maxNum) {
                maxIdx = i;
                maxNum = str[i];
            }
        }
        if(maxIdx == idx) {
            idx++;
        } else {
            k -= maxIdx - idx;
            str.splice(idx, maxIdx - idx);
        }
    }
    while(k) {
        str.pop();
        k--;
    }
    return str.join('');
}